crates/runtime/README.md

```diff
@@ -0,0 +1,662 @@
+system is particularly useful for passing data between different parts of the
+guest code or for storing intermediate results during computation.
+
+`VMLogic` also plays a vital role in managing the execution context. It holds
+the `VMContext` struct, which contains input data for the current execution.
+This allows the runtime to provide necessary contextual information to the guest
+code, such as function parameters or environmental data, in a controlled and
+standardised manner.
+
+Another critical aspect of `VMLogic` is its role in facilitating host function
+calls. Through the `host_functions()` method, it provides a set of functions
+that the guest code can invoke to interact with the host environment. These
+functions cover a wide range of operations, including memory and register
+manipulation, storage access, logging, event emission, and even external HTTP
+requests. By channeling all these interactions through `VMLogic`, the runtime
+can ensure that they are performed safely and within the defined limits.
+
+Throughout the execution of a Wasm module, `VMLogic` collects and manages logs
+and events emitted by the guest code. This feature is crucial for debugging,
+monitoring, and auditing purposes, allowing developers and system administrators
+to gain insights into the execution flow and any significant occurrences during
+runtime.
+
+When the execution of a Wasm function concludes, `VMLogic` is responsible for
+gathering all the results, logs, and events, and packaging them into an
+`Outcome` struct. This struct provides a comprehensive summary of the execution,
+including any return values or errors encountered during the process. By
+centralising this information, `VMLogic` ensures that the caller of the runtime
+receives a complete and consistent report of the execution results.
+
+### Host functions
+
+Host functions are implemented in the `imports.rs` file:
+
+  - Panic handling and UTF-8 string management
+  - Register operations (read, write, get length)
+  - Storage interactions
+  - Logging and event emission
+  - External HTTP requests (fetch)
+
+Host functions serve as the bridge between the sandboxed WebAssembly (Wasm)
+environment and the host system. These functions provide a carefully curated set
+of capabilities to the guest Wasm code, allowing it to interact with the outside
+world in a controlled and secure manner. The primary purpose of host functions
+is to extend the functionality of the Wasm module beyond its inherent
+limitations while maintaining strict control over its access to system resources
+and external services.
+
+At the core of the host function implementation is the `VMHostFunctions` struct,
+which encapsulates the runtime's logic and provides a safe interface for the
+guest code to interact with. This struct is created using self-references,
+allowing it to maintain a consistent view of the execution environment
+throughout the lifetime of a function call.
```

Comment:
Fair, I'm getting the feel some of these points are maybe too technical - I had written this to be technical documentation, for developers to read, but perhaps I've got the level wrong...? From a tech perspective I think knowing there's a self-reference is arguably useful, but otherwise not 🙂

Reply:
Well, I think from a tech perspective, you'll be reading the code, and in there the structure is annotated with `#[self_referencing]`

How I think about these docs is in the sense that some random person wants to get an idea why this component exists, and high-level - what it means to have it and how it functions.